In [24]:
#Import Necessary Library
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [25]:
#load the dataset
df=pd.read_csv('loan_approval_dataset.csv')

In [26]:
df.head()

loan_id   no_of_dependents      education  self_employed   income_annum  \
0        1                  2       Graduate             No        9600000   
1        2                  0   Not Graduate            Yes        4100000   
2        3                  3       Graduate             No        9100000   
3        4                  3       Graduate             No        8200000   
4        5                  5   Not Graduate            Yes        9800000   

    loan_amount   loan_term   cibil_score   residential_assets_value  \
0      29900000          12           778                    2400000   
1      12200000           8           417                    2700000   
2      29700000          20           506                    7100000   
3      30700000           8           467                   18200000   
4      24200000          20           382                   12400000   

    commercial_assets_value   luxury_assets_value   bank_asset_value  \
0                  17600000              22700000            8000000   
1                   2200000               8800000            3300000   
2                   4500000              33300000           12800000   
3                   3300000              23300000            7900000   
4                   8200000              29400000            5000000   

   loan_status  
0     Approved  
1     Rejected  
2     Rejected  
3     Rejected  
4     Rejected

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4269 entries, 0 to 4268
Data columns (total 13 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   loan_id                    4269 non-null   int64 
 1    no_of_dependents          4269 non-null   int64 
 2    education                 4269 non-null   object
 3    self_employed             4269 non-null   object
 4    income_annum              4269 non-null   int64 
 5    loan_amount               4269 non-null   int64 
 6    loan_term                 4269 non-null   int64 
 7    cibil_score               4269 non-null   int64 
 8    residential_assets_value  4269 non-null   int64 
 9    commercial_assets_value   4269 non-null   int64 
 10   luxury_assets_value       4269 non-null   int64 
 11   bank_asset_value          4269 non-null   int64 
 12   loan_status               4269 non-null   object
dtypes: int64(10), object(3)
memory usage: 433.7+ KB


In [28]:
df.describe()

loan_id   no_of_dependents   income_annum   loan_amount  \
count  4269.000000        4269.000000   4.269000e+03  4.269000e+03   
mean   2135.000000           2.498712   5.059124e+06  1.513345e+07   
std    1232.498479           1.695910   2.806840e+06  9.043363e+06   
min       1.000000           0.000000   2.000000e+05  3.000000e+05   
25%    1068.000000           1.000000   2.700000e+06  7.700000e+06   
50%    2135.000000           3.000000   5.100000e+06  1.450000e+07   
75%    3202.000000           4.000000   7.500000e+06  2.150000e+07   
max    4269.000000           5.000000   9.900000e+06  3.950000e+07   

         loan_term   cibil_score   residential_assets_value  \
count  4269.000000   4269.000000               4.269000e+03   
mean     10.900445    599.936051               7.472617e+06   
std       5.709187    172.430401               6.503637e+06   
min       2.000000    300.000000              -1.000000e+05   
25%       6.000000    453.000000               2.200000e+06   
50%      10.000000    600.000000               5.600000e+06   
75%      16.000000    748.000000               1.130000e+07   
max      20.000000    900.000000               2.910000e+07   

        commercial_assets_value   luxury_assets_value   bank_asset_value  
count              4.269000e+03          4.269000e+03       4.269000e+03  
mean               4.973155e+06          1.512631e+07       4.976692e+06  
std                4.388966e+06          9.103754e+06       3.250185e+06  
min                0.000000e+00          3.000000e+05       0.000000e+00  
25%                1.300000e+06          7.500000e+06       2.300000e+06  
50%                3.700000e+06          1.460000e+07       4.600000e+06  
75%                7.600000e+06          2.170000e+07       7.100000e+06  
max                1.940000e+07          3.920000e+07       1.470000e+07

In [29]:
df.shape

(4269, 13)

In [30]:
df.columns

Index(['loan_id', ' no_of_dependents', ' education', ' self_employed',
       ' income_annum', ' loan_amount', ' loan_term', ' cibil_score',
       ' residential_assets_value', ' commercial_assets_value',
       ' luxury_assets_value', ' bank_asset_value', ' loan_status'],
      dtype='object')

In [31]:
#checking the null values 
df.isnull().sum()

loan_id                      0
 no_of_dependents            0
 education                   0
 self_employed               0
 income_annum                0
 loan_amount                 0
 loan_term                   0
 cibil_score                 0
 residential_assets_value    0
 commercial_assets_value     0
 luxury_assets_value         0
 bank_asset_value            0
 loan_status                 0
dtype: int64

In [32]:
df.drop(['loan_id'],axis=1,inplace=True)

In [33]:
#Dividing datset into dependant and independent variable X & Y
x = df.drop([' loan_status'],axis=1)
y = df[' loan_status']

In [34]:
y=y.str.replace("Approved", "1")
y=y.str.replace("Rejected", "0")

In [35]:
print(x[' education'].unique())
print(x[' self_employed'].unique())

[' Graduate' ' Not Graduate']
[' No' ' Yes']


In [36]:
#We need to assign the binary to education and self_employed column to train the model
x[' education']=x[' education'].replace(' Graduate',"1")
x[' education']=x[' education'].replace(' Not Graduate',"0")
x[' self_employed']=x[' self_employed'].replace(' Yes',"1")
x[' self_employed']=x[' self_employed'].replace(' No',"0")

In [37]:
#Split data into Train and test Split
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.20,random_state=0)

In [38]:
x_train.shape

(3415, 11)

In [39]:
y_train.shape

(3415,)

In [40]:
#normaling the data 
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()

In [41]:
x_train= sc.fit_transform(x_train)
x_test= sc.transform(x_test)

In [42]:
#
#1.SVC
#2.Randomfirest
#3.xgboostclasifier
#4.logistic regression

In [43]:
#Now Importing necessary Models 
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
from sklearn import metrics

In [44]:
#intializing the Models
svc=SVC()
rf=RandomForestClassifier(n_estimators=50,criterion = 'entropy',random_state =10)
gbc=GradientBoostingClassifier()
lr=LogisticRegression()

In [45]:
#fititng the model on training data
for model in (svc, rf, gbc,lr):
    model.fit(x_train,y_train)
    y_pred=model.predict(x_train)
    print("Accuracy score of ",model.__class__.__name__,"=",100*metrics.accuracy_score(y_train,y_pred))

Accuracy score of  SVC = 95.51976573938506
Accuracy score of  RandomForestClassifier = 100.0
Accuracy score of  GradientBoostingClassifier = 99.64860907759882
Accuracy score of  LogisticRegression = 91.71303074670571


In [ ]:
#fititng the model on Testing data
for model in (svc, rf, gbc,lr):
    model.fit(x_train,y_train)
    y_pred=model.predict(x_test)
    print("Accuracy score of ",model.__class__.__name__,"=",100*metrics.accuracy_score(y_test,y_pred))

Accuracy score of  SVC = 94.14519906323186
Accuracy score of  RandomForestClassifier = 97.7751756440281
Accuracy score of  GradientBoostingClassifier = 98.47775175644028
Accuracy score of  LogisticRegression = 92.50585480093677


Conclusion

-RandomForestClassifier and GradientBoostingClassifier achieved the highest testing accuracies (97.78% and 98.48%, respectively).

-RandomForestClassifier, while having a perfect training accuracy (100%), still performed well on unseen data, suggesting robust generalization.

-Consider hyperparameter tuning for the chosen model, as it may further improve performance.

-Ensure data quality through preprocessing and feature engineering.